File: src/Object_Oriented_HackerRank_30DaysOfCode/linkedList/LinkedList.java
package Object_Oriented_HackerRank_30DaysOfCode.linkedList;

import com.sun.org.apache.bcel.internal.classfile.SourceFile;
import com.sun.org.apache.xpath.internal.SourceTree;

/**
 * Created by blaise on 8/8/17.
 */
public class LinkedList {
    Node head;
    int count;

    /**
     * @param head head/top of the linkedList
     */
    public LinkedList(Node head){
        this.head = head;
        this.count = 0;
    }

    /**
     * @param data data of the head of the linked list
     */
    public LinkedList(int data){
        this.head = new Node(data, null);
        this.count = 0;
    }

    public int getCount() {
        return count;
    }

    public Node getHead(){
        return this.head;
    }

    /**
     * @return length of the linked list
     */
    public int size(){
        return getCount() + 1;
    }

    /**
     *
     * @param data new node's data
     */
    public void insert(int data){
        if(this.getHead() != null) {
            Node temp = head;
            Node latest = new Node(data, null);
            while (temp.getNext() != null) {
                temp = temp.getNext();
            }
            temp.setNext(latest);
        }
        else
            head = new Node(data, null);

        this.count++;

    }

    /**
     * indexes start from 0 just like arrays
     * @param index of the node to be returned
     * @return data of the node at index=@param index
     */
    public int get(int index){

        Node temp = getHead();
        for(int i=1 ; i <= index; i++) {
            temp = temp.getNext();
        }
        return temp.getData();
    }

    /**
     * @return true if the list is empty
     */
    public boolean empty(){
        return this.head == null;
    }

    /**
     * removes the last element of the list
     */
    public void remove(){
        Node temp = getHead();
        for (int i = 0; i < this.getCount() - 1; i++) {
            temp = temp.getNext();
        }
        this.count--;
        temp.setNext(null);
    }
    public void printList(){
        Node temp = getHead();
        while(temp.next != null){
            System.out.print(temp.getData() + "-> ");
            temp = temp.getNext();
        }
        System.out.println(temp.getData());
    }
    public static void main(String [] args){
        LinkedList list = new LinkedList(1);
        for(int i=2; i<11; i++){
            list.insert(i);
        }
        list.printList();
        list.remove();
        list.printList();
        System.out.println(list.get(list.getCount()));
    }

}

File: src/data_structures/JavaScript/LinkedList.js
function LinkedList(){
  var head = null;
  var length = 0;
  
  var Node = function(element){
    this.element = element;
    this. next = null;
  }
  
  this.add = element =>
  {
    var node = new Node(element);
    if (head === null){
      head = node;
    }
    else {
      var last = head;
      while(last.next){
        last = last.next;
      }
      last.next = node;
    }
    length ++;
  }
  
  this.get = index =>{
    
    if(index < 0 ){
      return null;
    }
    
    var current = head;
    
    for(var i =0; i<length; i++){
      if(i === index){
        return current.element;
      }
      current = current.next;
    }
    return null;
  }
  
  this.remove = element =>{
    var current = head;
    var previous;
    if(head.element == element){
      head = head.next;
      length --;
    }
    else{
      while(current.element !== element){
        previous = current;
        current = current.next;
      }
      previous.next = current.next;
      length --;
    }
  }
  
  this.size = () => { return length}
  
  this.indexOf = element => {
    var current = head;
    for(var i = 0; i<length; i++){
      if(current.element !== element){
        current = current.next;
      }
      else{
        return i;
      }
    }
    return -1;
  }
  
  this.insert = (element, index) =>{
    
        
    if(index <0 || index >= length){
      return ;
    }
    
    var current = head;
    var node = new Node(element);
    var next;

    if(index === 0){
      node.next = head;
      head = node;
    }
    else{
     for(var i = 1; i<index; i++){
        current = current.next;
      }
    next = current.next;
    current.next = node;
    node.next = next;  
    }
    length ++;
  }
  
  this.removeAt = index =>{
    var current = head;
    if(index <0 || index >= length){
      return ;
    }
    if(index === 0){
      head = head.next;
    }
    else{
      for(var i= 1; i<index; i++){
        current = current.next;
      }
      current.next = current.next.next;
    }
    length --;
  }
  
  this.print = () =>{
    var str = '';
    var current = head;
    for(var i =0; i<length; i++){
      str += current.element +  ' -> ';
      current = current.next;
    }
    
    return str.substr(0,str.length-3);
  }
}


var list = new LinkedList();
for(var i =0; i<10; i++){
  list.add(i);
}

console.log(list.print());
console.log(list.size());
console.log(list.indexOf(3));
console.log(list.indexOf(10));
list.insert(-1,0);
console.log(list.print());
list.insert(12,2);
console.log(list.print());
console.log(list.size());
list.removeAt(0);
list.removeAt(1);
console.log(list.print());
console.log(list.get(3));
console.log(list.get(10));
